ethcore/src/client/client.rs

```diff
@@ -1794,56 +1894,60 @@ impl BlockChainClient for Client {
 		}
 	}
 
-	fn call_contract(&self, block_id: BlockId, address: Address, data: Bytes) -> Result<Bytes, String> {
-		let transaction = self.contract_call_tx(block_id, address, data);
-
-		self.call(&transaction, Default::default(), block_id)
-			.map_err(|e| format!("{:?}", e))
-			.map(|executed| {
-				executed.output
-			})
-	}
-
 	fn transact_contract(&self, address: Address, data: Bytes) -> Result<TransactionImportResult, EthcoreError> {
 		let transaction = Transaction {
-			nonce: self.latest_nonce(&self.miner.author()),
+			nonce: self.latest_nonce(&self.importer.miner.author()),
 			action: Action::Call(address),
-			gas: self.miner.gas_floor_target(),
-			gas_price: self.miner.sensible_gas_price(),
+			gas: self.importer.miner.gas_floor_target(),
+			gas_price: self.importer.miner.sensible_gas_price(),
 			value: U256::zero(),
 			data: data,
 		};
 		let chain_id = self.engine.signing_chain_id(&self.latest_env_info());
 		let signature = self.engine.sign(transaction.hash(chain_id))?;
 		let signed = SignedTransaction::new(transaction.with_signature(signature, chain_id))?;
-		self.miner.import_own_transaction(self, signed.into())
+		self.importer.miner.import_own_transaction(self, signed.into())
 	}
 
 	fn registrar_address(&self) -> Option<Address> {
 		self.registrar.lock().as_ref().map(|r| r.address)
 	}
 
-	fn registry_address(&self, name: String) -> Option<Address> {
-		self.registrar.lock().as_ref()
-			.and_then(|r| {
-				let dispatch = move |reg_addr, data| {
-					future::done(self.call_contract(BlockId::Latest, reg_addr, data))
-				};
-				r.get_address(dispatch, keccak(name.as_bytes()), "A".to_string()).wait().ok()
-			})
-			.and_then(|a| if a.is_zero() { None } else { Some(a) })
-	}
-
 	fn eip86_transition(&self) -> u64 {
 		self.engine().params().eip86_transition
 	}
 }
 
-impl MiningBlockChainClient for Client {
-	fn latest_schedule(&self) -> Schedule {
-		self.engine.schedule(self.latest_env_info().number)
+impl ReopenBlock for Client {
+	fn reopen_block(&self, block: ClosedBlock) -> OpenBlock {
+		let engine = &*self.engine;
+		let mut block = block.reopen(engine);
+		let max_uncles = engine.maximum_uncle_count(block.header().number());
+		if block.uncles().len() < max_uncles {
+			let chain = self.chain.read();
+			let h = chain.best_block_hash();
+			// Add new uncles
+			let uncles = chain
+				.find_uncle_hashes(&h, engine.maximum_uncle_age())
+				.unwrap_or_else(Vec::new);
+
+			for h in uncles {
+				if !block.uncles().iter().any(|header| header.hash() == h) {
+					let uncle = chain.block_header(&h).expect("find_uncle_hashes only returns hashes for existing headers; qed");
+					block.push_uncle(uncle).expect("pushing up to maximum_uncle_count;
+												push_uncle is not ok only if more than maximum_uncle_count is pushed;
+												so all push_uncle are Ok;
+												qed");
+					if block.uncles().len() >= max_uncles { break }
```

Comment:
what if the block had already maximum of uncles before reopening?

Reply:
This PR didn't introduced new logic here, code was moved from an existing `Client` impl. Still it's a good question.